File: swift/Sources/MudbaseShowcaseKanban/Support/ActivityText.swift
import Foundation

/// Renders one activity row as a plain, readable sentence for the feed. Ports
/// `../web/src/lib/activity-text.ts`'s `describeActivity` verbatim.
func describeActivity(_ entry: ActivityEntry) -> String {
    let who = entry.actorName.isEmpty ? "Someone" : entry.actorName
    switch entry.action {
    case .createdCard:
        return "\(who) created card \"\(entry.cardTitle ?? "Untitled")\" in \(entry.toList ?? "a list")"
    case .moved:
        if let fromList = entry.fromList, let toList = entry.toList, fromList == toList {
            return "\(who) reordered \"\(entry.cardTitle ?? "a card")\" within \(toList)"
        }
        return "\(who) moved \"\(entry.cardTitle ?? "a card")\" from \(entry.fromList ?? "?") to \(entry.toList ?? "?")"
    case .deletedCard:
        return "\(who) deleted card \"\(entry.cardTitle ?? "Untitled")\" from \(entry.fromList ?? "a list")"
    case .createdList:
        return "\(who) created list \"\(entry.toList ?? "Untitled")\""
    case .renamedList:
        return "\(who) renamed list \"\(entry.fromList ?? "?")\" to \"\(entry.toList ?? "?")\""
    case .deletedList:
        return "\(who) deleted list \"\(entry.fromList ?? "Untitled")\""
    }
}

/// SF Symbol per action, mirroring the icon-per-action mapping in
/// `../web/src/components/activity/ActivityItem.tsx`.
func activityIconName(_ action: ActivityAction) -> String {
    switch action {
    case .createdCard: return "plus.circle"
    case .moved: return "arrow.left.arrow.right"
    case .deletedCard: return "trash"
    case .createdList: return "rectangle.stack.badge.plus"
    case .renamedList: return "arrow.triangle.2.circlepath"
    case .deletedList: return "rectangle.stack.badge.minus"
    }
}

File: go/internal/server/activity_text.go
package server

import (
	"fmt"

	"github.com/mudbase/mudbase-showcase-kanban/go/internal/models"
)

// describeActivity renders one activity row as a plain, readable sentence for the feed. Ported
// from the reference web app's src/lib/activity-text.ts (describeActivity).
func describeActivity(entry models.Activity) string {
	who := entry.ActorName
	if who == "" {
		who = "Someone"
	}

	cardTitle := entry.CardTitle
	if cardTitle == "" {
		cardTitle = "a card"
	}

	switch entry.Action {
	case models.ActionCreatedCard:
		toList := entry.ToList
		if toList == "" {
			toList = "a list"
		}
		return fmt.Sprintf("%s created card %q in %s", who, orUntitled(entry.CardTitle), toList)
	case models.ActionMoved:
		if entry.FromList != "" && entry.FromList == entry.ToList {
			return fmt.Sprintf("%s reordered %q within %s", who, cardTitle, entry.ToList)
		}
		from := entry.FromList
		if from == "" {
			from = "?"
		}
		to := entry.ToList
		if to == "" {
			to = "?"
		}
		return fmt.Sprintf("%s moved %q from %s to %s", who, cardTitle, from, to)
	case models.ActionDeletedCard:
		fromList := entry.FromList
		if fromList == "" {
			fromList = "a list"
		}
		return fmt.Sprintf("%s deleted card %q from %s", who, orUntitled(entry.CardTitle), fromList)
	case models.ActionCreatedList:
		return fmt.Sprintf("%s created list %q", who, orUntitled(entry.ToList))
	case models.ActionRenamedList:
		from := entry.FromList
		if from == "" {
			from = "?"
		}
		to := entry.ToList
		if to == "" {
			to = "?"
		}
		return fmt.Sprintf("%s renamed list %q to %q", who, from, to)
	case models.ActionDeletedList:
		return fmt.Sprintf("%s deleted list %q", who, orUntitled(entry.FromList))
	default:
		return fmt.Sprintf("%s did something on the board", who)
	}
}

func orUntitled(s string) string {
	if s == "" {
		return "Untitled"
	}
	return s
}
